File: ios/App/Shared/SleepActionSender.swift
import Foundation

/// Sender en søvnlog-handling til backend og holder styr på de handlinger, der
/// endnu ikke er nået frem.
///
/// Filen kompileres ind i **både** app-targetet og widget-extensionen. Knappen i
/// Live Activity'en køres som en `LiveActivityIntent`, og iOS kører den i appens
/// egen proces — men extensionen skal kunne kompilere intentet for overhovedet
/// at kunne vise knappen. Derfor må der ikke bruges noget her, som kun findes i
/// appen: ingen UIKit, ingen Capacitor.
enum SleepActionSender {

    // MARK: - Konstanter

    static let appId = "699f47a86e7e0a874d1159ed"
    static let defaultEndpoint =
        "https://lalatoto.base44.app/api/apps/699f47a86e7e0a874d1159ed/functions/nativeSleepAction"

    /// Capacitor Preferences gemmer i UserDefaults med præfikset "CapacitorStorage."
    static let tokenKey = "CapacitorStorage.cns_native_token"
    static let endpointKey = "CapacitorStorage.cns_native_endpoint"

    /// Vores egen kø over handlinger, der endnu ikke er nået frem.
    static let queueKey = "cns_pending_sleep_actions"

    /// Backend mapper 'awake' til mark_awake og 'sleeping' til mark_sleeping.
    static let actionAwake = "awake"
    static let actionSleeping = "sleeping"
    static let actionEnd = "end"

    /// Køen holdes kort. Bliver en handling ved med at blive afvist, ryger den
    /// ud, når der er kommet 50 nyere ind foran den.
    private static let maxQueueLength = 50

    private static let urlSession: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 20
        config.waitsForConnectivity = false
        return URLSession(configuration: config)
    }()

    // MARK: - Udførelse

    /// Gemmer handlingen i køen og forsøger straks at sende den. Lykkes det,
    /// fjernes den igen fra køen.
    ///
    /// Der gemmes ALTID først, så trykket er registreret, også hvis appen bliver
    /// lukket ned midt i netværkskaldet.
    ///
    /// Rækkefølgen er også det, der gør trykket sikkert på en låst telefon.
    /// UserDefaults ligger som udgangspunkt under beskyttelsesklassen
    /// «tilgængelig efter første oplåsning siden genstart», og der er ikke sat
    /// en strengere klasse noget sted i projektet, så tokenet kan normalt læses
    /// fra låseskærmen. Kan det alligevel ikke — for eksempel hvis telefonen er
    /// genstartet og aldrig låst op siden — fejler kun afsendelsen, mens
    /// handlingen bliver liggende i køen og går igennem ved næste appstart.
    static func perform(action: String,
                        sessionId: String?,
                        completion: @escaping (Bool) -> Void) {
        let stamp = iso8601(Date())
        enqueue(action: action, sessionId: sessionId, at: stamp)

        send(action: action, sessionId: sessionId, at: stamp) { settled in
            if settled {
                dequeue(at: stamp)
                NSLog("[CNS-NATIVE] handling sendt: %@", action)
            } else {
                NSLog("[CNS-NATIVE] handling bliver i køen: %@", action)
            }
            completion(settled)
        }
    }

    /// Sender én handling. `completion(true)` betyder "sagen er afsluttet — tag
    /// den ud af køen", ikke nødvendigvis at den lykkedes.
    static func send(action: String,
                     sessionId: String?,
                     at stamp: String,
                     completion: @escaping (Bool) -> Void) {

        let defaults = UserDefaults.standard
        guard let token = defaults.string(forKey: tokenKey), !token.isEmpty else {
            // Tokenet skrives af appen efter login. Er det her endnu ikke, skal
            // handlingen blive i køen og prøves igen ved næste appstart.
            NSLog("[CNS-NATIVE] intet token i Preferences — handlingen bliver i køen")
            completion(false)
            return
        }

        let endpointString = defaults.string(forKey: endpointKey) ?? defaultEndpoint
        guard let url = URL(string: endpointString) else {
            completion(false)
            return
        }

        var body: [String: Any] = ["token": token, "action": action, "at": stamp]
        if let sessionId = sessionId { body["session_id"] = sessionId }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(appId, forHTTPHeaderField: "X-App-Id")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        urlSession.dataTask(with: request) { _, response, error in
            if let error = error {
                NSLog("[CNS-NATIVE] netværksfejl: %@", error.localizedDescription)
                completion(false)
                return
            }
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            NSLog("[CNS-NATIVE] svar: %d", status)

            // Kun to slags svar er endelige:
            //
            //   2xx — handlingen er udført.
            //   404 — der er ingen aktiv session at udføre den på. Den kommer
            //         aldrig tilbage, så det nytter ikke at prøve igen.
            //
            // Alt andet bliver i køen. Særligt 401: det betyder som regel bare,
            // at tokenet endnu ikke var skrevet til Preferences, da trykket
            // skete — for eksempel ved allerførste tryk efter login. Smider man
            // handlingen væk der, mister brugeren trykket permanent.
            if (200...299).contains(status) || status == 404 {
                completion(true)
            } else {
                completion(false)
            }
        }.resume()
    }

    // MARK: - Kø

    static func enqueue(action: String, sessionId: String?, at stamp: String) {
        var queue = UserDefaults.standard.array(forKey: queueKey) as? [[String: Any]] ?? []
        var item: [String: Any] = ["action": action, "at": stamp]
        if let sessionId = sessionId { item["session_id"] = sessionId }
        queue.append(item)
        if queue.count > maxQueueLength { queue = Array(queue.suffix(maxQueueLength)) }
        UserDefaults.standard.set(queue, forKey: queueKey)
    }

    static func dequeue(at stamp: String) {
        let queue = UserDefaults.standard.array(forKey: queueKey) as? [[String: Any]] ?? []
        let remaining = queue.filter { ($0["at"] as? String) != stamp }
        UserDefaults.standard.set(remaining, forKey: queueKey)
    }

    /// Sender ventende handlinger igen. Kaldes ved appstart.
    static func flushQueue() {
        let queue = UserDefaults.standard.array(forKey: queueKey) as? [[String: Any]] ?? []
        guard !queue.isEmpty else { return }
        NSLog("[CNS-NATIVE] tømmer kø: %d ventende", queue.count)

        for item in queue {
            guard let action = item["action"] as? String,
                  let stamp = item["at"] as? String else { continue }
            let sessionId = item["session_id"] as? String
            send(action: action, sessionId: sessionId, at: stamp) { settled in
                if settled { dequeue(at: stamp) }
            }
        }
    }

    // MARK: - Hjælpere

    static func iso8601(_ date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: date)
    }
}

File: android/app/src/main/java/com/base699f47a86e7e0a874d1159ed/app/SleepActionSender.java
package com.base699f47a86e7e0a874d1159ed.app;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * Sender en søvnlog-handling til backend og holder styr på de handlinger, der
 * endnu ikke er nået frem.
 *
 * Modstykket til SleepActionSender.swift på iOS. Samme endepunkt, samme
 * handlingsnavne, samme regler for hvornår en handling er endelig, og samme kø
 * med oprindeligt tidsstempel.
 */
final class SleepActionSender {

    private static final String TAG = "CNS-NATIVE";

    static final String APP_ID = "699f47a86e7e0a874d1159ed";
    static final String DEFAULT_ENDPOINT =
        "https://lalatoto.base44.app/api/apps/699f47a86e7e0a874d1159ed/functions/nativeSleepAction";

    /** Capacitor Preferences gemmer på Android i SharedPreferences-filen "CapacitorStorage". */
    private static final String CAPACITOR_PREFS = "CapacitorStorage";
    private static final String TOKEN_KEY = "cns_native_token";
    private static final String ENDPOINT_KEY = "cns_native_endpoint";

    /** Vores egen kø over handlinger, der endnu ikke er nået frem. */
    private static final String QUEUE_PREFS = "cns_sleep_queue";
    private static final String QUEUE_KEY = "cns_pending_sleep_actions";

    /** Backend mapper 'awake' til mark_awake og 'sleeping' til mark_sleeping. */
    static final String ACTION_AWAKE = "awake";
    static final String ACTION_SLEEPING = "sleeping";
    static final String ACTION_END = "end";

    /**
     * Køen holdes kort. Bliver en handling ved med at blive afvist, ryger den ud,
     * når der er kommet 50 nyere ind foran den.
     */
    private static final int MAX_QUEUE_LENGTH = 50;

    private SleepActionSender() {}

    // ---------------------------------------------------------------- Udførelse

    /**
     * Gemmer handlingen i køen og forsøger straks at sende den. Lykkes det,
     * fjernes den igen fra køen.
     *
     * Der gemmes ALTID først, så trykket er registreret, også hvis processen
     * bliver slået ihjel midt i netværkskaldet.
     *
     * Kalder netværket direkte og må derfor kun køres fra en baggrundstråd.
     */
    static void perform(Context context, String action, String sessionId) {
        String stamp = iso8601(new Date());
        enqueue(context, action, sessionId, stamp);

        boolean settled = send(context, action, sessionId, stamp);
        if (settled) {
            dequeue(context, stamp);
            Log.d(TAG, "handling sendt: " + action);
        } else {
            Log.d(TAG, "handling bliver i køen: " + action);
        }
    }

    /**
     * Sender én handling. Returnerer true når sagen er afsluttet — altså "tag den
     * ud af køen" — ikke nødvendigvis at den lykkedes.
     */
    static boolean send(Context context, String action, String sessionId, String stamp) {
        SharedPreferences prefs =
            context.getSharedPreferences(CAPACITOR_PREFS, Context.MODE_PRIVATE);

        String token = prefs.getString(TOKEN_KEY, null);
        if (token == null || token.isEmpty()) {
            // Tokenet skrives af appen efter login. Er det her endnu ikke, skal
            // handlingen blive i køen og prøves igen ved næste appstart.
            Log.d(TAG, "intet token i Preferences — handlingen bliver i køen");
            return false;
        }

        String endpoint = prefs.getString(ENDPOINT_KEY, DEFAULT_ENDPOINT);

        HttpURLConnection connection = null;
        try {
            JSONObject body = new JSONObject();
            body.put("token", token);
            body.put("action", action);
            body.put("at", stamp);
            if (sessionId != null && !sessionId.isEmpty()) {
                body.put("session_id", sessionId);
            }

            connection = (HttpURLConnection) new URL(endpoint).openConnection();
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(20000);
            connection.setReadTimeout(20000);
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json");
            connection.setRequestProperty("X-App-Id", APP_ID);

            try (OutputStream out = connection.getOutputStream()) {
                out.write(body.toString().getBytes(StandardCharsets.UTF_8));
            }

            int status = connection.getResponseCode();
            Log.d(TAG, "svar: " + status);

            // Kun to slags svar er endelige:
            //
            //   2xx — handlingen er udført.
            //   404 — der er ingen aktiv session at udføre den på. Den kommer
            //         aldrig tilbage, så det nytter ikke at prøve igen.
            //
            // Alt andet bliver i køen. Særligt 401: det betyder som regel bare,
            // at tokenet endnu ikke var skrevet til Preferences, da trykket
            // skete. Smider man handlingen væk der, mister brugeren trykket.
            return (status >= 200 && status <= 299) || status == 404;
        } catch (Exception e) {
            Log.w(TAG, "netværksfejl: " + e.getMessage());
            return false;
        } finally {
            if (connection != null) connection.disconnect();
        }
    }

    // ---------------------------------------------------------------------- Kø

    private static SharedPreferences queuePrefs(Context context) {
        return context.getSharedPreferences(QUEUE_PREFS, Context.MODE_PRIVATE);
    }

    private static JSONArray readQueue(Context context) {
        String raw = queuePrefs(context).getString(QUEUE_KEY, "[]");
        try {
            return new JSONArray(raw);
        } catch (Exception e) {
            return new JSONArray();
        }
    }

    private static void writeQueue(Context context, JSONArray queue) {
        queuePrefs(context).edit().putString(QUEUE_KEY, queue.toString()).apply();
    }

    static void enqueue(Context context, String action, String sessionId, String stamp) {
        try {
            JSONArray queue = readQueue(context);

            JSONObject item = new JSONObject();
            item.put("action", action);
            item.put("at", stamp);
            if (sessionId != null && !sessionId.isEmpty()) {
                item.put("session_id", sessionId);
            }
            queue.put(item);

            while (queue.length() > MAX_QUEUE_LENGTH) {
                queue.remove(0);
            }
            writeQueue(context, queue);
        } catch (Exception e) {
            Log.w(TAG, "kunne ikke lægge handling i kø: " + e.getMessage());
        }
    }

    static void dequeue(Context context, String stamp) {
        JSONArray queue = readQueue(context);
        JSONArray remaining = new JSONArray();
        for (int i = 0; i < queue.length(); i++) {
            JSONObject item = queue.optJSONObject(i);
            if (item == null) continue;
            if (!stamp.equals(item.optString("at"))) {
                remaining.put(item);
            }
        }
        writeQueue(context, remaining);
    }

    /** Sender ventende handlinger igen. Kaldes ved appstart. */
    static void flushQueue(Context context) {
        JSONArray queue = readQueue(context);
        if (queue.length() == 0) return;
        Log.d(TAG, "tømmer kø: " + queue.length() + " ventende");

        for (int i = 0; i < queue.length(); i++) {
            JSONObject item = queue.optJSONObject(i);
            if (item == null) continue;

            String action = item.optString("action", null);
            String stamp = item.optString("at", null);
            if (action == null || stamp == null) continue;

            String sessionId = item.optString("session_id", null);
            if (send(context, action, sessionId, stamp)) {
                dequeue(context, stamp);
            }
        }
    }

    /** Som flushQueue, men på egen tråd — netværk må ikke røre hovedtråden. */
    static void flushQueueAsync(final Context context) {
        final Context appContext = context.getApplicationContext();
        new Thread(new Runnable() {
            @Override
            public void run() {
                flushQueue(appContext);
            }
        }).start();
    }

    // --------------------------------------------------------------- Hjælpere

    /** Samme format som iOS sender: ISO 8601 i UTC uden brøkdele af sekunder. */
    static String iso8601(Date date) {
        SimpleDateFormat formatter =
            new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));
        return formatter.format(date);
    }
}
